plugins/SlicerT/SlicerTView.cpp

```diff
@@ -0,0 +1,202 @@
+	, m_resetButton(this)
+	, m_midiExportButton(this)
+	, m_wf(248, 128, instrument, this)
+{
+	// window settings
+	setAcceptDrops(true);
+	setAutoFillBackground(true);
+
+	// render background
+	QPalette pal;
+	pal.setBrush(backgroundRole(), PLUGIN_NAME::getIconPixmap("artwork"));
+	setPalette(pal);
+
+	// move editor and seeker
+	m_wf.move(2, 6);
+
+	// snap combo box
+	m_snapSetting.setGeometry(185, 200, 55, ComboBox::DEFAULT_HEIGHT);
+	m_snapSetting.setToolTip(tr("Set slice snapping for detection"));
+	m_snapSetting.setModel(&m_slicerTParent->m_sliceSnap);
+
+	// sync toggle
+	m_syncToggle.move(135, 187);
+	m_syncToggle.setToolTip(tr("Enable BPM sync"));
+	m_syncToggle.setModel(&m_slicerTParent->m_enableSync);
+
+	// bpm spin box
+	m_bpmBox.move(130, 201);
+	m_bpmBox.setToolTip(tr("Original sample BPM"));
+	/* m_bpmBox.setLabel(tr("BPM")); */
+	m_bpmBox.setModel(&m_slicerTParent->m_originalBPM);
+
+	// threshold knob
+	m_noteThresholdKnob = createStyledKnob();
+	m_noteThresholdKnob->move(10, 197);
+	m_noteThresholdKnob->setToolTip(tr("Threshold used for slicing"));
+	/* m_noteThresholdKnob->setLabel(tr("Threshold")); */
+	m_noteThresholdKnob->setModel(&m_slicerTParent->m_noteThreshold);
+
+	// fadeout knob
+	m_fadeOutKnob = createStyledKnob();
+	m_fadeOutKnob->move(64, 197);
+	m_fadeOutKnob->setToolTip(tr("Fade Out for notes"));
+	/* m_fadeOutKnob->setLabel(tr("Fade Out")); */
+	m_fadeOutKnob->setModel(&m_slicerTParent->m_fadeOutFrames);
+
+	// midi copy button
+	m_midiExportButton.move(199, 150);
+	m_midiExportButton.setIcon(PLUGIN_NAME::getIconPixmap("copyMidi"));
+	m_midiExportButton.setToolTip(tr("Copy midi pattern to clipboard"));
+	connect(&m_midiExportButton, &PixmapButton::clicked, this, &SlicerTView::exportMidi);
+
+	// slice reset button
+	m_resetButton.move(18, 150);
+	m_resetButton.setIcon(PLUGIN_NAME::getIconPixmap("resetSlices"));
+	m_resetButton.setToolTip(tr("Reset Slices"));
+	connect(&m_resetButton, &PixmapButton::clicked, m_slicerTParent, &SlicerT::updateSlices);
+}
+
+// style knob, defined in data/themes/default/style.css#L949
+Knob* SlicerTView::createStyledKnob()
+{
+	Knob* newKnob = new Knob(KnobType::Styled, this);
+	newKnob->setFixedSize(50, 40);
+	newKnob->setCenterPointX(24.0);
+	newKnob->setCenterPointY(15.0);
+	return newKnob;
+}
+
+// copied from piano roll
+void SlicerTView::exportMidi()
+{
+	using namespace Clipboard;
+	if (m_slicerTParent->m_originalSample.frames() < 2048) { return; }
+
+	DataFile dataFile(DataFile::Type::ClipboardData);
+	QDomElement note_list = dataFile.createElement("note-list");
+	dataFile.content().appendChild(note_list);
+
+	std::vector<Note> notes = m_slicerTParent->getMidi();
+	if (notes.size() == 0) { return; }
+
+	TimePos start_pos(notes.front().pos().getBar(), 0);
+	for (Note note : notes)
+	{
+		Note clip_note(note);
+		clip_note.setPos(clip_note.pos(start_pos));
+		clip_note.saveState(dataFile, note_list);
+	}
```

Comment:
Why make two copies? First you copy the `Note` into `note`, then the same note into `clip_note`. Maybe just take a reference in the ranged for loop itself?

Reply:
No idea, I copied the code from the piano roll, as stated in the comment above. Just using `note` seems to work fine, so that's done.

Unrelated, but while I was testing if everything works, I found that lmms crashes if you: place a few notes in piano roll, select all these notes, control+rightArrow, alt+leftArrow. Just wanted to say it here, but if it's something important maybe an issue should be opened.

Reply:
> Unrelated, but while I was testing if everything works, I found that lmms crashes if you: place a few notes in piano roll, select all these notes, control+rightArrow, alt+leftArrow. Just wanted to say it here, but if it's something important maybe an issue should be opened.

Great find!